Block toolbar: split switcher from mover and simplify styles.

ZebulanStanphill opened the pull request:
## Description
Following the discussion starting [here](https://github.com/WordPress/gutenberg/issues/21935#issuecomment-654405787), this PR simplifies the block toolbar styles so that the switcher and movers are shown in separate toolbar groups, with the standard spacing between the two. They were already technically in separate toolbar groups, but the styles obscured this to try and show them closer together.

To ensure that the parent block selector button and block outline only show when the block icon is hovered, I had to move some elements around a bit. As it turns out, this allowed me to remove even more special-case CSS.

I initially tried to create a PR that simply increased the space between the switcher and mover, following [this comment](https://github.com/WordPress/gutenberg/issues/21935#issuecomment-657902142) by @diegohaz. However, this proved to be really, really complicated to try and implement. As it turns out, going the next step and just using the standard styles was a **drastically** simpler solution.

As the accessibility report found the combined switcher/mover area to be a regression, accessibility-wise, I'm tagging this with the "Backport to WP Core" label in the hope that it will be included in WP 5.5, as [suggested](https://github.com/WordPress/gutenberg/pull/23760#issuecomment-658170074) by @afercia.

## Screenshots

![image](https://user-images.githubusercontent.com/19592990/87616844-ab29c400-c6db-11ea-9c58-7d2867fa9a30.png)
![image](https://user-images.githubusercontent.com/19592990/87616853-b7158600-c6db-11ea-8490-d86c64228f2a.png)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ZebulanStanphill committed:
Block toolbar: simplify switcher/mover styles.

ZebulanStanphill commented:
It turns out the bug I fixed in https://github.com/WordPress/gutenberg/pull/23971/commits/a669dc3648fb1f861864becad6ced90f06c9fd64 was actually present in `master`, too. I've created #23980 to fix the bug in a separate PR, assuming this one doesn't get merged first.

youknowriad commented:
There's still a lot of discussions about that and different opinions. I removed the backport label but if it an agreement is reached and it makes it before 5.5, we'll add it back.

ZebulanStanphill commented:
Rebased since #23980 got merged.

ZebulanStanphill commented:
The current end-to-end test failure looks unrelated. Going to rebase to restart the tests and see if it passes. Just thought I'd note the failing test right now.
```
FAIL packages/e2e-tests/specs/editor/various/adding-patterns.test.js (5.766s)
  ● adding blocks › should insert a block pattern

    expect(received).toMatchSnapshot()

    Snapshot name: `adding blocks should insert a block pattern 1`

    - Snapshot  - 2
    + Received  + 2

      <!-- wp:buttons {"align":"center"} -->
      <div class="wp-block-buttons aligncenter"><!-- wp:button {"borderRadius":0,"backgroundColor":"very-dark-gray"} -->
    - <div class="wp-block-button"><a class="wp-block-button__link has-very-dark-gray-background-color has-background no-border-radius">Get started</a></div>
    + <div class="wp-block-button"><a class="wp-block-button__link has-very-dark-gray-background-color has-background no-border-radius">Button One</a></div>
      <!-- /wp:button -->

      <!-- wp:button {"borderRadius":0,"textColor":"very-dark-gray","className":"is-style-outline"} -->
    - <div class="wp-block-button is-style-outline"><a class="wp-block-button__link has-very-dark-gray-color has-text-color no-border-radius">Learn more</a></div>
    + <div class="wp-block-button is-style-outline"><a class="wp-block-button__link has-very-dark-gray-color has-text-color no-border-radius">Button Two</a></div>
      <!-- /wp:button --></div>
      <!-- /wp:buttons -->

      18 | 		await insertPattern( 'Two buttons' );
      19 | 
    > 20 | 		expect( await getEditedPostContent() ).toMatchSnapshot();
         | 		                                       ^
      21 | 	} );
      22 | } );
      23 | 

      at Object.<anonymous> (specs/editor/various/adding-patterns.test.js:20:42)
          at runMicrotasks (<anonymous>)
```

ZebulanStanphill commented:
The test is still failing, but it's also failing on other PRs, so it's definitely not caused by this one.

ZebulanStanphill commented:
The failing test issue was fixed by #24004. Rebased yet again, and now all tests are passing. :smile:

ZebulanStanphill committed:
Move mover out of block-switcher-container

Only show parent block selector button and block outline when hovering the block icon.

ZebulanStanphill commented:
Rebased again.

joedolson commented:
The accessibility team would definitely like to see this happen in WP 5.5; it would help a lot in mitigating some of the challenges with this part of the control.

ZebulanStanphill committed:
Reduce style specificity and add missing comment.

shaunandrews commented:
This makes the movers a little less visually obvious; With the added divider between the block icon and the movers, the context feels different. My instinct tells me that the movers should be somewhat _special_ in comparison to the rest of the buttons in the block toolbar.

Looking back at @diegohaz's suggestion, I like the extra spacing instead of a more explicit divider; The space helps connect the movers to the block in question, and feels overall visually simpler.

<img width="362" alt="image" src="https://user-images.githubusercontent.com/191598/89426341-0e6dab80-d708-11ea-9df1-c119707b8241.png">

I wonder if this PR could accomplish this by hiding the divider between the block icon and the movers.

paaljoachim commented:
I am pinging a few folks who added a comment to the issue, so they have a chance to take a look at this PR.
@pablohoneyhoney @jasmussen @ellatrix

ZebulanStanphill commented:
Semantically, though, the controls _should_ be divided by a line because they're in separate `ToolbarGroup`s. Also, the dividing line helps make it clear that the movers look are not part of the switcher. The a11y team has pointed out that the current design's movers look like dropdown indicators for a `<select>`.

This also ties in another issue (that hopefully someone else can solve): the movers use the same icons as the accordion panels, `<select>` dropdowns (not just in general but in our own UI style), and inline formatting tool overflow, which is very bad for comprehension since the same icons are being used in different ways across the UI.

Another issue is that the mover buttons being stacked in the first place makes it a lot more difficult to where one starts and ends. The icons are so close together they appear like a single button.

There's yet another issue tying into all this: we need to have a visible drag handle. Showing a drag-hand cursor when hovering the mover buttons was deemed bad for a11y (all buttons should use the standard button cursor), so it was reverted. But now there's no indicator of any way to drag blocks. There ought to be a clearly defined drag handle somewhere. And to pull that off, the movers will ultimately have to be split apart into 2 full-size buttons... probably with a drag handle in-between.

These _all_ have to be addressed in order to undo a11y regressions that WP 5.5 is introducing.

I realize that fitting 3 full-size buttons into the block toolbar is asking a lot. But I think that speaks to a larger issue: we need a way to show controls more contextually. We can't always show the block movement controls, because there simply isn't enough space to do that. In general, there's simply too many toolbar controls to show all at once.

I've been thinking about recent explorations like #24021. **Assuming it can be done accessibly**, I think we should apply this concept to the movement controls: by default, a movement controls button would appear in the toolbar. Clicking it would swap the currently visible controls with all the movement-related ones. This would allow us to not only show up, down, and a drag handle, but also additional controls like "move to top" or "move to bottom".

Anyway, this PR was intended to be a compromise to improve the situation for now by removing some of the weird special-casing that makes the current situation bad for a11y. Trying to step back a bit closer to the current design in `master` makes the CSS really complicated again or ends up looking inconsistent with the rest of the toolbar, while barely addressing the usability/a11y problems.

If this PR isn't a good enough compromise for everyone, I suggest we look into the full-toolbar-replacement idea and try to get it done in an accessible way in time for WP 5.6.

pablohoneyhoney commented:
In my opinion, we can move affirmatively with this. Though, I think we’ll need to revise soon how the dividers and grouping work in the anatomy of the block toolbar and its different cases. Because this change introduces complexity in the hierarchy of the controls, their taxonomy, and also establishes a busier toolbar.

jasmussen reviewed (approved):
Thanks for the PR! This is what I see:

![movers](https://user-images.githubusercontent.com/1204802/89527678-07b76500-d7ea-11ea-9dfe-728fe5230383.gif)

High level, this seems like a good enough compromise that keeps intact the benefits of having the mover control in the toolbar (no lateral overlap in nested contexts, consistency across mobile and desktop, works identically for vertical and lateral movement). Closes #23760. 

Keeping Pablo's comments in mind, let's try it.

ntsekouras reviewed (approved):
LGTM 👍  Thanks!

paaljoachim commented:
Btw can we make the up and down arrow buttons cover the full space they are in? As it would make the hit area bigger.

ZebulanStanphill commented:
Thanks, everyone! :tada:

mapk commented:
It's great to see this happened! 🎉  Thanks @ZebulanStanphill
